Add some support for detecting false aliasing in AOTAutograd

bdhirsh opened the pull request:
This is a partial fix for https://github.com/pytorch/pytorch/issues/106457. In the examples with the shampoo optimizer that i ran, they were enough to remove the parameter aliasing in shampoo.

I added some new logic for detecting if two inputs have overlapping memory in specific cases: if they're both 2D tensors with stride 1. In that case (the case for shampoo), I try to compute a bunch of contiguous intervals on the two tensors, and check if any of the intervals overlap. In theory this is slow, since if our two tensors are e.g. of size (256, N), we'll need to create 256 intervals to check for overlap on. This seems... probably fine, since I think we do more egregious things in the compile stack to cause slowness. Open to suggestions though!

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106461
* #106460

bdhirsh committed:
Add some support for detecting false aliasing in AOTAutograd

[ghstack-poisoned]

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,52 @@ def same_dtype_views(view1, view2):
+    # We can compute a list of contiguous intervals and check if any of them intersect
+    if x.dim() == 2 and y.dim() == 2 and x.stride()[1] == 1 and y.stride()[1] == 1:
+        # Compute the (start, end) range of contiguous memory of each region of the tensor
+        x_ranges = [
```
my code is under the assumption that `x.stride(0) == y.stride(0)` for sure! But that seems to be the case you care about anyways?

bdhirsh replied:
Talked offline - agreed to add the faster check you mentioned to the C++ code (which does not have that check today, since it bails out if either tensor is noncontiguous)

bdhirsh committed:
Update on "Add some support for detecting false aliasing in AOTAutograd"

This is a partial fix for https://github.com/pytorch/pytorch/issues/106457. In the examples with the shampoo optimizer that i ran, they were enough to remove the parameter aliasing in shampoo.

I added some new logic for detecting if two inputs have overlapping memory in specific cases: if they're both 2D tensors with stride 1. In that case (the case for shampoo), I try to compute a bunch of contiguous intervals on the two tensors, and check if any of the intervals overlap. In theory this is slow, since if our two tensors are e.g. of size (256, N), we'll need to create 256 intervals to check for overlap on. This seems... probably fine, since I think we do more egregious things in the compile stack to cause slowness. Open to suggestions though!




[ghstack-poisoned]

ezyang commented:
@albanD you think you can do full review for this one?

albanD commented:
Yes I can do a full review on this.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,52 @@ def same_dtype_views(view1, view2):
+def compute_overlapping_inputs(fwd_inputs, aliased_input_indices):
+    actual_aliased_indices = set()
+    for j in range(len(aliased_input_indices)):
+        for i in range(j):
```
So the idea here is that this "check if there is any real overlapping memory" happens in two stages:

Stage 1: We take all 200 of the graph inputs, and bucket them based on their `.untyped_storage()`. For each bucket, we keep track of a set of "what are (the indices of) the graph inputs that all share a storage. Say that out of our 200 graph inputs, one bucket is [37, 38, 39]

Stage 2: We take each of those buckets of indices, and figure out if any of them have "true overlapping memory". So `aliased_input_indices` here would refer to `[37, 38, 39]`, and we check pair-wise if any pair of tensors is overlapping.

Two reasons that I set it up this way:

(1) Stage 1 already exists, so it was easier to add a separate independent stage in for the extra overlapping memory checks

(2) In theory, doing this "check pair-wise if any pair of tensors has overlapping memory" can be pretty expensive, since (a) I saw that shampoo had 10-20 params that all shared a storage (and we have to do O(n^2) overlapping memory checks), and (b) the checks themselves can be slow (like this overlapping-interval logic that I added).

I agree the two stage logic is a bit complicated, but lmk if this sounds reasonable to you!

albanD replied:
That sounds good, Thanks for the details!

bdhirsh commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -731,6 +732,14 @@ void initDispatchBindings(PyObject* module) {
     return at::functionalization::impl::commit_update(a);
   });
+  m.def(
+      "_are_definitely_non_overlapping",
```
hmm, I couldn't find a binding for it

bdhirsh committed:
Update on "Add some support for detecting false aliasing in AOTAutograd"

This is a partial fix for https://github.com/pytorch/pytorch/issues/106457. In the examples with the shampoo optimizer that i ran, they were enough to remove the parameter aliasing in shampoo.

I added some new logic for detecting if two inputs have overlapping memory in specific cases: if they're both 2D tensors with stride 1. In that case (the case for shampoo), I try to compute a bunch of contiguous intervals on the two tensors, and check if any of the intervals overlap. In theory this is slow, since if our two tensors are e.g. of size (256, N), we'll need to create 256 intervals to check for overlap on. This seems... probably fine, since I think we do more egregious things in the compile stack to cause slowness. Open to suggestions though!




[ghstack-poisoned]

bdhirsh commented:
@albanD this is ready for another round. I ended up:

(1) Not re-using the C++ MemOverlap code at all, instead of trying to improve it (if I had more time I would try, but shampoo folks would like this fix in soon)

(2) Beefed up the AOTAutograd python mem overlap code in two ways:

  (a) added your faster/better overlap check
  (b) added a quick short-circuit that returns false when the two tensors obviously overlap (easy to tell when both are contiguous)

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,74 @@ def same_dtype_views(view1, view2):
+
+    # This cases is needed for the shampoo optimizer.
+    # We can compute a list of contiguous intervals and check if any of them intersect
+    if x.dim() == 2 and y.dim() == 2 and x.stride()[1] == 1 and y.stride()[1] == 1:
```
nit: this is not c++ `x.stride(1)` will be faster

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,74 @@ def same_dtype_views(view1, view2):
+        # Simple case first: both inputs have the same outer stride
+        if x.stride(0) == y.stride(0):
+            if x.storage_offset() + x.size(0) > y.storage_offset():
+                return True
```
False?

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,74 @@ def same_dtype_views(view1, view2):
+            if x.storage_offset() + x.size(0) > y.storage_offset():
+                return True
+            if y.storage_offset() + y.size(0) > x.stride(0):
+                return True
```
False?

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,74 @@ def same_dtype_views(view1, view2):
+        else:
+            return intervals_intersect(xs[1:], ys)
+
+    # This cases is needed for the shampoo optimizer.
```
Not sure how shampoo helps here?
Example of two Tensors would be more useful here I think

bdhirsh replied:
> Not sure how shampoo helps here?

Whoops thanks - the comment about shampoo was meant for lower down (the bit handling intervals)

> Example of two Tensors would be more useful here I think

You mean adding a comment that shows a small example of what this case looks like? Definitely, I can add one.

albanD replied:
Yes showing two small Tensors that would trigger this particular lack of overlap pattern.

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,74 @@ def same_dtype_views(view1, view2):
+            if y.storage_offset() + y.size(0) > x.stride(0):
+                return True
+
+        # Slightly more complicated case, to account for 2d tensors with different outer strides.
```
What is an example where you need this use case?

albanD reviewed (commented):
Since your function has a bunch of special cases, I would expect to see one test for each of these cases.

bdhirsh committed:
Update on "Add some support for detecting false aliasing in AOTAutograd"

This is a partial fix for https://github.com/pytorch/pytorch/issues/106457. In the examples with the shampoo optimizer that i ran, they were enough to remove the parameter aliasing in shampoo.

I added some new logic for detecting if two inputs have overlapping memory in specific cases: if they're both 2D tensors with stride 1. In that case (the case for shampoo), I try to compute a bunch of contiguous intervals on the two tensors, and check if any of the intervals overlap. In theory this is slow, since if our two tensors are e.g. of size (256, N), we'll need to create 256 intervals to check for overlap on. This seems... probably fine, since I think we do more egregious things in the compile stack to cause slowness. Open to suggestions though!




[ghstack-poisoned]

bdhirsh commented:
Updated PR. Main changes include:

- The "check for false aliasing" code is now fully in python - I didn't bother trying to re-use and improve the C++ code (also didn't want to have to worry about symint-ifying it now to get it to work with dynamic shapes)
- I changed the algorithm so the checks are all constant time, and confirmed on my shampoo example internally that we still remove the aliasing
- more comprehensive testing

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1648,6 +1648,83 @@ def same_dtype_views(view1, view2):
+                # definitely overlaps (row 0 of y overlaps with row 0 of x)
+                # Example:
+                #   x: size=(4, 4), stride=(8, 1), offset=0
+                #   y: size=(4, 4), stride=(8, 1), offset=3
```
nit: I would mention that you get these as `x = base[:4]` and `y = base[3:]`. That makes it easier to visualize for me.

bdhirsh replied:
Hmm... is there actually a nice set of torch view ops that represents the strides and storage_offsets in my example? I don't think `base[:4]` and `base[3:]` will reflect the fact that x and y are non-contiguous (because they have a large outer stride)

bdhirsh replied:
yep you're right, that's exactly it. will update

albanD reviewed (approved):
Sounds good to me!

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1110,7 +1110,7 @@ def inp_callable(req_grad):
             x = base.add(1)
             inp1 = x[0]
-            inp2 = x[1]
+            inp2 = x[0]
```
the original purpose of this test was to test out the synthetic base behavior... but AOTAutograd is now smart enough to realize that these old inputs don't actually alias (they just share a storage) 😛. I just had to update a couple of the synthetic base tests to *actually* overlap memory.

albanD replied:
Ho ok! Sounds good!

bdhirsh committed:
Update on "Add some support for detecting false aliasing in AOTAutograd"

This is a partial fix for https://github.com/pytorch/pytorch/issues/106457. In the examples with the shampoo optimizer that i ran, they were enough to remove the parameter aliasing in shampoo.

I added some new logic for detecting if two inputs have overlapping memory in specific cases: if they're both 2D tensors with stride 1. In that case (the case for shampoo), I try to compute a bunch of contiguous intervals on the two tensors, and check if any of the intervals overlap. In theory this is slow, since if our two tensors are e.g. of size (256, N), we'll need to create 256 intervals to check for overlap on. This seems... probably fine, since I think we do more egregious things in the compile stack to cause slowness. Open to suggestions though!




[ghstack-poisoned]

bdhirsh committed:
Update on "Add some support for detecting false aliasing in AOTAutograd"

This is a partial fix for https://github.com/pytorch/pytorch/issues/106457. In the examples with the shampoo optimizer that i ran, they were enough to remove the parameter aliasing in shampoo.

I added some new logic for detecting if two inputs have overlapping memory in specific cases: if they're both 2D tensors with stride 1. In that case (the case for shampoo), I try to compute a bunch of contiguous intervals on the two tensors, and check if any of the intervals overlap. In theory this is slow, since if our two tensors are e.g. of size (256, N), we'll need to create 256 intervals to check for overlap on. This seems... probably fine, since I think we do more egregious things in the compile stack to cause slowness. Open to suggestions though!




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot rebase

bdhirsh committed:
Update on "Add some support for detecting false aliasing in AOTAutograd"

This is a partial fix for https://github.com/pytorch/pytorch/issues/106457. In the examples with the shampoo optimizer that i ran, they were enough to remove the parameter aliasing in shampoo.

I added some new logic for detecting if two inputs have overlapping memory in specific cases: if they're both 2D tensors with stride 1. In that case (the case for shampoo), I try to compute a bunch of contiguous intervals on the two tensors, and check if any of the intervals overlap. In theory this is slow, since if our two tensors are e.g. of size (256, N), we'll need to create 256 intervals to check for overlap on. This seems... probably fine, since I think we do more egregious things in the compile stack to cause slowness. Open to suggestions though!




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
